cli/lsp/completions.rs

```diff
@@ -486,45 +503,168 @@ fn parse_npm_specifier_version_index(specifier: &str) -> Option<usize> {
           .find('@')
           .filter(|idx2| !bare_specifier[idx..][1..*idx2].is_empty())
           .filter(|idx2| !bare_specifier[idx..][1..*idx2].contains('/'))
-          .map(|idx2| 4 + idx + idx2)
+          .map(|idx2| idx + idx2)
       })
   } else {
     bare_specifier
       .find('@')
       .filter(|idx| !bare_specifier[1..*idx].is_empty())
       .filter(|idx| !bare_specifier[1..*idx].contains('/'))
-      .map(|idx| 4 + idx)
   }
 }
 
+async fn get_jsr_completions(
+  referrer: &ModuleSpecifier,
+  specifier: &str,
+  range: &lsp::Range,
+  jsr_search_api: &impl PackageSearchApi,
+  jsr_resolver: &JsrResolver,
+) -> Option<Vec<lsp::CompletionItem>> {
+  // First try to match `jsr:some-package@some-version/<export-to-complete>`.
+  if let Ok(req_ref) = JsrPackageReqReference::from_str(specifier) {
+    let sub_path = req_ref.sub_path();
+    if sub_path.is_some() || specifier.ends_with('/') {
+      let export_prefix = sub_path.unwrap_or("");
+      let req = req_ref.req();
+      let nv = jsr_resolver.resolve_req(req);
+      let nv = nv.or_else(|| PackageNv::from_str(&req.to_string()).ok())?;
+      let exports = jsr_search_api.exports(&nv).await.ok()?;
+      let items = exports
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, export)| {
+          if export == "." {
+            return None;
+          }
+          let export = export.strip_prefix("./").unwrap_or(export.as_str());
+          if !export.starts_with(export_prefix) {
+            return None;
+          }
+          let specifier = format!("jsr:{}/{}", req_ref.req(), export);
+          let command = Some(lsp::Command {
+            title: "".to_string(),
+            command: "deno.cache".to_string(),
+            arguments: Some(vec![json!([&specifier]), json!(referrer)]),
+          });
```

Comment:
Does this cause a cache when accepting the completion? The one annoying thing about that might be is that it could add a bunch of lockfile entries. So for example, you accept for `1.2.3` then you change it to `@^1.2` then cache and now you have two entries (since specifiers in files aren't tracked by the lockfile).

Reply:
I've now made it so that caching on accepting completions always uses the global cache (no `vendor`) and no lockfile. Both are still used when caching in other deliberate cases and when checking if a file is cached. Of course the lockfile will always be respected for execution.

Reply:
Nice. Good idea!